src/workerd/api/headers.h

```diff
@@ -157,16 +162,67 @@ class Headers final: public jsg::Object {
 
   JSG_SERIALIZABLE(rpc::SerializationTag::HEADERS);
 
-  void visitForMemoryInfo(jsg::MemoryTracker& tracker) const {
-    for (const auto& entry : headers) {
-      tracker.trackField(entry.first, entry.second);
-    }
-  }
+  void visitForMemoryInfo(jsg::MemoryTracker& tracker) const;
 
 private:
-  struct Header {
-    jsg::ByteString key;   // lower-cased name
-    jsg::ByteString name;
+  // The header key can be stored either as a common header ID (uint)
+  class UncommonHeaderKey final {
+   public:
+    UncommonHeaderKey(UncommonHeaderKey&&) = default;
+    UncommonHeaderKey& operator=(UncommonHeaderKey&&) = default;
+    KJ_DISALLOW_COPY(UncommonHeaderKey);
+
+    bool equals(const UncommonHeaderKey& other) const;
+
+    // Will incur a string copy of otherName internally to perform a case-insensitive comparison.
+    // In the event that such a comparison is necessary to be performed multiple times, it is
+    // recommended to construct an UncommonHeaderKey instance once and use that for comparisons.
+    bool equals(kj::StringPtr otherName) const;
+
+    uint hashCode() const { return hash; }
+    kj::StringPtr getName() const { return name; }
+
+    JSG_MEMORY_INFO(UncommonHeaderKey) {
+      tracker.trackField("name", name);
+    }
+
+    UncommonHeaderKey clone() const {
+      return UncommonHeaderKey(kj::str(name), hash);
+    }
+
+   private:
+    // The name is expected to be stored in lower-case form, but we do not
+    // enforce that actually within the struct. It is the responsibility of
+    // the caller to ensure it is appropriately lower-cased.
+    kj::String name;
+    kj::uint hash;
+
+    // Critically, because of hash collisions, we must still compare the full string
+    // to determine equality but we can use the hash code to avoid unnecessary string
+    // comparisons.
+    UncommonHeaderKey(kj::String name);
+    UncommonHeaderKey(kj::StringPtr name);
+    UncommonHeaderKey(kj::String name, kj::uint hash): name(kj::mv(name)), hash(hash) {}
+    friend class Headers;
+  };
+
+  // A header is identified by either a common header ID or an uncommon header name.
+  // The header key name is always identifed in lower-case form, while the original
+  // casing is preserved in the actual Header struct to support case-preserving display.
+  using HeaderKey = kj::OneOf<uint, UncommonHeaderKey>;
+
+  static HeaderKey getHeaderKeyFor(kj::StringPtr name);
+  static bool headerKeyEquals(const HeaderKey& a, const HeaderKey& b);
+  static HeaderKey cloneHeaderKey(const HeaderKey& key);
+  static bool isSetCookie(const HeaderKey& key);
```

Comment:
What's the purpose of this? Are we going to add every combination as a static method here?

Reply:
See the usage. set-cookie is special.